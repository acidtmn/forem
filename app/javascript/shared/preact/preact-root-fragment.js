/**
 * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
 *
 * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
 * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
 * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
 */
export function createRootFragment(parent, replaceNode) {
  replaceNode = [].concat(replaceNode);
  var s = replaceNode[replaceNode.length-1].nextSibling;
  function insert(c, r) { parent.insertBefore(c, r || s); }
  return parent.__k = {
    nodeType: 1,
    parentNode: parent,
    firstChild: replaceNode[0],
    childNodes: replaceNode,
    insertBefore: insert,
    appendChild: insert,
    removeChild: function(c) { parent.removeChild(c); }
  };
}