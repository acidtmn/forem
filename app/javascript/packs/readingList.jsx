import { h } from 'preact';
import { ReadingList } from '../readingList/readingList';
import { render } from '@utilities/preact';

function loadElement() {
  const root = document.getElementById('reading-list');
  if (root) {
    render(
      <ReadingList availableTags={[]} statusView={root.dataset.view} />,
      root,
    );
  }
}

window.InstantClick.on('change', () => {
  loadElement();
});

loadElement();
