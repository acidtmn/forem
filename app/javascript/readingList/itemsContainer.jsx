import { h } from 'preact';
import { PropTypes } from 'preact-compat';

import { ItemListItem } from '../src/components/ItemList/ItemListItem';
import { ItemListItemArchiveButton } from '../src/components/ItemList/ItemListItemArchiveButton';
import { ItemListLoadMoreButton } from '../src/components/ItemList/ItemListLoadMoreButton';

const ItemListItemContainer = ({
  items,
  archiveButtonLabel,
  toggleArchiveStatus,
}) => {
  return items.map(item => {
    return (
      <ItemListItem item={item}>
        <ItemListItemArchiveButton
          text={archiveButtonLabel}
          onClick={e => toggleArchiveStatus(e, item)}
        />
      </ItemListItem>
    );
  });
};

const FilterText = ({ selectedTags, query, value }) => {
  return (
    <h1>
      {selectedTags.length === 0 && query.length === 0
        ? value
        : 'Nothing with this filter ðŸ¤”'}
    </h1>
  );
};

const EmptyItems = ({ itemsLoaded, selectedTags, query }) => {
  if (itemsLoaded && isStatusViewValid) {
    return (
      <div className="items-empty">
        <FilterText
          selectedTags={selectedTags}
          query={query}
          value="Your Reading List is Lonely"
        />
        <h3>
          Hit the
          <span className="highlight">SAVE</span>
          or
          <span className="highlight">
            Bookmark
            <span role="img" aria-label="Bookmark">
              ðŸ”–
            </span>
          </span>
          to start your Collection
        </h3>
      </div>
    );
  }

  return (
    <div className="items-empty">
      <FilterText
        selectedTags={selectedTags}
        query={query}
        value="Your Archive List is Lonely"
      />
    </div>
  );
};

const ItemsContainer = ({
  itemsLoaded,
  items,
  archiveButtonLabel,
  toggleArchiveStatus,
  selectedTags,
  query,
  showLoadMoreButton,
  loadNextPage,
  totalCount,
  isStatusViewValid,
}) => {
  return (
    <div className="items-container">
      <div className={`results ${itemsLoaded ? 'results--loaded' : ''}`}>
        <div className="results-header">
          {isStatusViewValid ? 'Reading List' : 'Archive'}
          {` (${totalCount > 0 ? totalCount : 'empty'})`}
        </div>
        <div>
          {items.length > 0 ? (
            <ItemListItemContainer
              items={items}
              archiveButtonLabel={archiveButtonLabel}
              toggleArchiveStatus={toggleArchiveStatus}
            />
          ) : (
            <EmptyItems
              itemsLoaded={itemsLoaded}
              selectedTags={selectedTags}
              query={query}
            />
          )}
        </div>
      </div>
      <ItemListLoadMoreButton
        show={showLoadMoreButton}
        onClick={loadNextPage}
      />
    </div>
  );
};

FilterText.propTypes = {
  selectedTags: PropTypes.arrayOf(PropTypes.string).isRequired,
  value: PropTypes.string.isRequired,
  query: PropTypes.arrayOf(PropTypes.string).isRequired,
};

EmptyItems.propTypes = {};

ItemsContainer.propTypes = {};

export default ItemsContainer;
