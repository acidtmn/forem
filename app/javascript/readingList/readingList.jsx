import { h, Component } from 'preact';
import { PropTypes } from 'preact-compat';
import debounce from 'lodash.debounce';
import setupAlgoliaIndex from '../src/utils/algolia';

import { ItemListLoadMoreButton } from '../src/components/ItemList/ItemListLoadMoreButton';
import { ItemListTags } from '../src/components/ItemList/ItemListTags';

const STATUS_VIEW_VALID = 'valid';
const STATUS_VIEW_ARCHIVED = 'archived';

export class ReadingList extends Component {
  constructor(props) {
    super(props);

    this.handleTyping = debounce(this.handleTyping.bind(this), 300, {
      leading: true,
    });

    const { availableTags, statusView } = this.props;
    this.state = {
      query: '',
      index: null,

      page: 0,
      hitsPerPage: 1,
      totalCount: 0,

      items: [],
      itemsLoaded: false,

      availableTags,
      selectedTags: [],

      showLoadMoreButton: false,

      archiving: false,
      statusView,
    };
  }

  componentDidMount() {
    const index = setupAlgoliaIndex({
      containerId: 'reading-list',
      indexName: 'SecuredReactions',
    });

    // get default result set from Algolia
    const { hitsPerPage, statusView } = this.state;
    index
      .search('', { hitsPerPage, filters: `status:${statusView}` })
      .then(content => {
        this.setState({
          items: content.hits,
          totalCount: content.nbHits,
          index,
          itemsLoaded: true,
          showLoadMoreButton: content.hits.length === hitsPerPage,
        });
      });
  }

  handleTyping = e => {
    const query = e.target.value;
    const { selectedTags, statusView } = this.state;
    this.listSearch(query, selectedTags, statusView);
  };

  toggleTag = (e, tag) => {
    e.preventDefault();
    const { query, selectedTags, statusView } = this.state;
    const newTags = selectedTags;
    if (newTags.indexOf(tag) === -1) {
      newTags.push(tag);
    } else {
      newTags.splice(newTags.indexOf(tag), 1);
    }
    this.setState({ selectedTags: newTags, page: 0, items: [] });
    this.listSearch(query, newTags, statusView);
  };

  toggleStatusView = e => {
    e.preventDefault();
    const { statusView, query, selectedTags } = this.state;
    if (statusView === STATUS_VIEW_VALID) {
      this.setState({ statusView: STATUS_VIEW_ARCHIVED, page: 0, items: [] });
      this.listSearch(query, selectedTags, STATUS_VIEW_ARCHIVED);
      window.history.replaceState(null, null, '/readinglist/archive');
    } else {
      this.setState({ statusView: STATUS_VIEW_VALID, page: 0, items: [] });
      this.listSearch(query, selectedTags, STATUS_VIEW_VALID);
      window.history.replaceState(null, null, '/readinglist');
    }
  };

  archive = (e, item) => {
    e.preventDefault();
    const { statusView, items, totalCount } = this.state;
    const t = this;
    window.fetch(`/reading_list_items/${item.id}`, {
      method: 'PUT',
      headers: {
        'X-CSRF-Token': window.csrfToken,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ current_status: statusView }),
      credentials: 'same-origin',
    });
    const newItems = items;
    newItems.splice(newItems.indexOf(item), 1);
    t.setState({
      archiving: true,
      items: newItems,
      totalCount: totalCount - 1,
    });
    setTimeout(() => {
      t.setState({ archiving: false });
    }, 1800);
  };

  loadNextPage = () => {
    const { statusView, query, selectedTags, page } = this.state;
    const isLoadMore = true;
    this.setState({ page: page + 1 });
    this.listSearch(query, selectedTags, statusView, isLoadMore);
  };

  listSearch(query, tags, statusView) {
    const { index, hitsPerPage, page, items } = this.state;
    const filters = { page, hitsPerPage, filters: `status:${statusView}` };
    if (tags.length > 0) {
      filters.tagFilters = tags;
    }
    index.search(query, filters).then(content => {
      const allItems = [...items, ...content.hits];

      this.setState({
        query,
        items: allItems,
        totalCount: content.nbHits,
        showLoadMoreButton: content.hits.length === hitsPerPage,
      });
    });
  }

  render() {
    const {
      archiving,
      availableTags,
      itemsLoaded,
      query,
      items,
      selectedTags,
      statusView,
      showLoadMoreButton,
      totalCount,
    } = this.state;
    let allItems = items.map(item => (
      <div className="readinglist-item-wrapper">
        <a className="readinglist-item" href={item.searchable_reactable_path}>
          <div className="readinglist-item-title">
            {item.searchable_reactable_title}
          </div>
          <div className="readinglist-item-details">
            <a
              className="readinglist-item-user"
              href={`/${item.reactable_user.username}`}
            >
              <img
                src={item.reactable_user.profile_image_90}
                alt="Profile Pic"
              />
              {item.reactable_user.name}
ãƒ»
              {item.reactable_published_date}
ãƒ»
              {item.reading_time}
              {' '}
min readãƒ»
            </a>
            <span className="readinglist-item-tag-collection">
              {item.reactable_tags.map(tag => (
                <a className="readinglist-item-tag" href={`/t/${tag}`}>
                  #
                  {tag}
                </a>
              ))}
            </span>
          </div>
        </a>
        <button
          className="readinglist-archive-butt"
          onClick={e => this.archive(e, item)}
          type="button"
        >
          {statusView === 'valid' ? 'archive' : 'unarchive'}
        </button>
      </div>
    ));
    if (items.length === 0 && itemsLoaded) {
      if (statusView === 'valid') {
        allItems = (
          <div className="readinglist-empty">
            <h1>
              {selectedTags.length === 0 && query.length === 0
                ? 'Your Reading List is Lonely'
                : 'Nothing with this filter ðŸ¤”'}
            </h1>
            <h3>
              Hit the
              <span className="highlight">SAVE</span>
              or
              <span className="highlight">
                Bookmark
                <span role="img" aria-label="Bookmark">
                  ðŸ”–
                </span>
              </span>
              to start your Collection
            </h3>
          </div>
        );
      } else {
        allItems = (
          <div className="readinglist-empty">
            <h1>
              {selectedTags.length === 0 && query.length === 0
                ? 'Your Archive List is Lonely'
                : 'Nothing with this filter ðŸ¤”'}
            </h1>
          </div>
        );
      }
    }

    const snackBar = archiving ? (
      <div className="snackbar">
        {statusView === STATUS_VIEW_VALID ? 'Archiving' : 'Unarchiving'}
        (async)
      </div>
    ) : (
      ''
    );

    return (
      <div className="home readinglist-home">
        <div className="side-bar">
          <div className="widget readinglist-filters">
            <input onKeyUp={this.handleTyping} placeHolder="search your list" />
            <ItemListTags
              availableTags={availableTags}
              selectedTags={selectedTags}
              onClick={this.toggleTag}
            />
            <div className="readinglist-view-toggle">
              <a
                href="/readinglist/archive"
                onClick={e => this.toggleStatusView(e)}
                data-no-instant
              >
                {statusView === STATUS_VIEW_VALID
                  ? 'View Archive'
                  : 'View Reading List'}
              </a>
            </div>
          </div>
        </div>
        <div className="items-container">
          <div
            className={`readinglist-results ${
              itemsLoaded ? 'readinglist-results--loaded' : ''
            }`}
          >
            <div className="readinglist-results-header">
              {statusView === STATUS_VIEW_VALID ? 'Reading List' : 'Archive'}
              {` (${totalReadingList > 0 ? totalReadingList : 'empty'})`}
            </div>
            <div>{allItems}</div>
          </div>
          <ItemListLoadMoreButton
            show={showLoadMoreButton}
            onClick={this.loadNextPage}
          />
        </div>
        {snackBar}
      </div>
    );
  }
}

ReadingList.defaultProps = {
  statusView: STATUS_VIEW_VALID,
};

ReadingList.propTypes = {
  availableTags: PropTypes.arrayOf(PropTypes.string).isRequired,
  statusView: PropTypes.oneOf([STATUS_VIEW_VALID, STATUS_VIEW_ARCHIVED]),
};
