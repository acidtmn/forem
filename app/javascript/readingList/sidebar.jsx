import { h } from 'preact';

import { ItemListTags } from '../src/components/ItemList/ItemListTags';

const SideBar = ({
  onSearchBoxType,
  isStatusViewValid,
  selectedTags,
  clearSelectedTags,
  availableTags,
  toggleTag,
  toggleStatusView,
}) => {
  return (
    <div className="side-bar">
      <div className="widget filters">
        <input onKeyUp={onSearchBoxType} placeHolder="search your list" />
        <div className="filters-header">
          <h4 className="filters-header-text">my tags</h4>
          {Boolean(selectedTags.length) && (
            <a
              className="filters-header-action"
              href={
                isStatusViewValid
                  ? READING_LIST_PATH
                  : READING_LIST_ARCHIVE_PATH
              }
              onClick={clearSelectedTags}
              data-no-instant
            >
              clear all
            </a>
          )}
        </div>
        <ItemListTags
          availableTags={availableTags}
          selectedTags={selectedTags}
          onClick={toggleTag}
        />

        <div className="status-view-toggle">
          <a
            href={READING_LIST_ARCHIVE_PATH}
            onClick={e => toggleStatusView(e)}
            data-no-instant
          >
            {isStatusViewValid ? 'View Archive' : 'View Reading List'}
          </a>
        </div>
      </div>
    </div>
  );
};

export default SideBar;
