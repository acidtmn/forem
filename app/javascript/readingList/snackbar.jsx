import { h } from 'preact';

const SnackBar = ({ archiving, isStatusViewValid }) => {
  if (archiving) {
    return (
      <div className="snackbar">
        {isStatusViewValid() ? 'Archiving...' : 'Unarchiving...'}
      </div>
    );
  }
};

export default SnackBar;
