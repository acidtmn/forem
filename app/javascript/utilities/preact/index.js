import { render as preactRender } from 'preact';
import { unmountComponentAtNode } from 'preact/compat';

/**
 * Renders a Preact component into the given parent DOM element.
 *
 * Note: This function is calling Preact's render under the hood along with
 * a call to InstantClick's change event that explicitly unmounts the component rendered.
 *
 * @param {ComponentChild} vnode
 * @param {Element | Document | ShadowRoot | DocumentFragment} parent
 * @param {Element | Text} replaceNode
 */
export function render(vnode, parent, replaceNode) {
  InstantClick.on('change', () => {
    parent && unmountComponentAtNode(parent);
  });

  preactRender(vnode, parent, replaceNode);
}
