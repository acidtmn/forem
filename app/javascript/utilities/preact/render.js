import { render as preactRender } from 'preact';
import { unmountComponentAtNode } from 'preact/compat';

/**
 * Renders a Preact component into the given parent DOM element.
 *
 * Note: This function is calling Preact's render under the hood along with
 * a call to InstantClick's change event that explicitly unmounts the component rendered.
 *
 * @param {object} vnode The virtual DOM node representing a component.
 * @param {Element | Document | ShadowRoot | DocumentFragment} parent The DOM element where the component will render.
 * @param {Element | Text} replaceNode An optional DOM element that must of the given parent.
 * Instead of inferring where to start rendering, it will update or replace the passed element using Preact's diffing algorithm.
 */
export function render(vnode, parent, replaceNode) {
  InstantClick.on('change', () => {
    // We need to explicitly unmount a Preact component since we are using InstantClick. Failing to do so
    // would leave references to past components.
    parent && unmountComponentAtNode(parent);
  });

  preactRender(vnode, parent, replaceNode);
}
