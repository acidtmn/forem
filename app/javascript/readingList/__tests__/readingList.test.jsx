import { h } from 'preact';
import { render } from '@testing-library/preact';
import { ReadingList } from '../readingList';

describe('<ReadingList />', () => {
  it('renders all the elements', () => {
    const { getByPlaceholderText, getByText } = render(<ReadingList availableTags={['discuss']} />);

    expect(getByPlaceholderText('search your list')).toBeTruthy();
    expect(getByText('#discuss')).toBeTruthy();
    expect(getByText('View Archive')).toBeTruthy();
    expect(getByText('Your Archive List is Lonely')).toBeTruthy();
    expect(getByText('Reading List (empty)')).toBeTruthy();
  });
});
