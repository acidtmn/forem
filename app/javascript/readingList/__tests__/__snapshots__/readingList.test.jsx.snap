// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReadingList /> renders properly 1`] = `
<div
  class="home readinglist-home"
>
  <div
    class="side-bar"
  >
    <div
      class="widget readinglist-filters"
    >
      <input
        onKeyUp={[Function]}
        placeHolder="search your list"
      />
      <div
        class="readinglist-tags"
      >
        <a
          class="readinglist-tag "
          data-no-instant={true}
          href="/t/discuss"
          onClick={[Function]}
        >
          #discuss
        </a>
      </div>
      <div
        class="readinglist-view-toggle"
      >
        <a
          data-no-instant={true}
          href="/readinglist/archive"
          onClick={[Function]}
        >
          View Archive
        </a>
      </div>
    </div>
  </div>
  <div
    class="readinglist-result-container"
  >
    <div
      class="readinglist-results "
    >
      <div
        class="readinglist-results-header"
      >
        Reading List (0)
      </div>
      <div />
    </div>
    <div
      class="loadmore-wrapper"
    />
  </div>
</div>
`;
